### PR TITLE
Add tests for exercising the LLVM backend with Unicode input 

### LINKED ISSUE #385: Add test for throwing Kore with unicode to the LLVM backend
Currently, we are testing throwing unicode characters to:

- `kast`: frontend CLI tool, and
- `kore-rpc`: (tests skipped) for execution.

We should also test the LLVM backend.

### PR DESCRIPTION
**Pull request by tothtamas28**
Fixes #385 

Adds tests for using `krun` and the LLVM bindings.

### CODE REVIEW

File: src/tests/integration/test_string.py [hunk 1]
```diff
@@ -128,33 +171,52 @@ def test_cli_rule_to_kast(definition_dir: Path, text: str) -> None:
...
+
+    # When
+    actual = krun.run_kore_term(kore)
+
+    # Then
+    assert actual == expected
 
+
+@pytest.mark.parametrize('text', TEST_DATA, ids=TEST_DATA)
+def test_bindings(runtime: ModuleType, text: str) -> None:
```
Comment (tothtamas28) [inline src/tests/integration/test_string.py:188]: Note that this test is not disabled, so things seem to break only when going through a textual KORE interface. (Or if the string is manipulated during rewrite.)

File: src/tests/integration/test_string.py [hunk 2]
```diff
@@ -128,33 +171,52 @@ def test_cli_rule_to_kast(definition_dir: Path, text: str) -> None:
...
-                App("Lbl'-LT-'k'-GT-'", (), (kseq,)),
-                App("Lbl'-LT-'generatedCounter'-GT-'", (), (int_dv(0),)),
-                App("Lbl'-LT-'s'-GT-'", (), (s,)),
-            ),
-        )
+def test_krun(haskell_dir: Path, text: str) -> None:
+    # Given
+    kore = kore_config(text, '')
+    expected = kore_config(None, text)
+    krun = KRun(haskell_dir)
```
Comment (tothtamas28) [inline src/tests/integration/test_string.py:178]: If you want to see this break:

```suggestion
def test_krun(definition_dir: Path, text: str) -> None:
    # Given
    kore = kore_config(text, '')
    expected = kore_config(None, text)
    krun = KRun(definition_dir)
```

File: src/tests/integration/test_string.py (hunk 1 above)
Comment (Scott-Guest) [reply src/tests/integration/test_string.py:188]: This behavior is expected as far as I can tell.

The Unicode issues with the LLVM backend that I'm aware of are
1. `\xhh` is treated as a raw byte `0xhh`, but Pyk, the Java frontend, and the Haskell backend treat it as `U+00hh`
2.  Unicode characters are UTF-8 encoded to be able to fit into a `char[]`, causing problems when
    1. manipulating strings (because the operations are not Unicode-aware)
    2. mixing Unicode characters with `\xhh` escapes

We never run into (1) or (2i) here because the escape sequences are only relevant during textual parsing / unparsing, but `llvm_to_kore` and `kore_to_llvm` just directly convert AST to AST (with pybind UTF-8 encoding and decoding at the boundary). We never run into 2ii because we don't manipulate the string.

File: src/tests/integration/test_string.py (hunk 2 above)
Comment (Scott-Guest) [reply src/tests/integration/test_string.py:178]: Can we do something like
```
@pytest.mark.parametrize('text', TEST_DATA, ids=TEST_DATA)
def test_krun(definition_dir: Path, llvm_dir: Path, text: str) -> None:
    if definition_dir == llvm_dir:
        # https://github.com/runtimeverification/k/issues/3344
        pytest.skip()
```
to document that this is failing test which is intentionally skipped?

File: src/tests/integration/test_string.py [hunk 3]
```diff
@@ -6,39 +6,82 @@
...
-KOMPILE_MAIN_FILE = 'k-files/string-rewrite.k'
-KOMPILE_MAIN_MODULE = 'STRING-REWRITE'
-
 TEST_DATA: Final = (
     'hello',
     '\n',
+    '\x01',
     '武天老師',
     '🙂',
 )
```
Comment (Scott-Guest) [inline src/tests/integration/test_string.py:37]: Let's add a few more test cases to ensure we hit [all of the UTF-8 encoding ranges](https://en.wikipedia.org/wiki/UTF-8#Encoding): £ (U+00A3) and 𐍈 (U+10348)

File: src/tests/integration/test_string.py [hunk 4]
```diff
@@ -128,33 +171,52 @@ def test_cli_rule_to_kast(definition_dir: Path, text: str) -> None:
...
         pytest.skip()
 
     # Given
-    init = config(string_dv(text), string_dv(''))
+    kore = kore_config(text, '')
+    expected = kore_config(None, text)
 
     # When
-    with KoreServer(definition_dir, KOMPILE_MAIN_MODULE) as server:
+    with KoreServer(haskell_dir, KOMPILE_MAIN_MODULE) as server:
```
Comment (Scott-Guest) [inline src/tests/integration/test_string.py:217]: Might just be my lack of familiarity with `KoreServer`, but why exclude the `llvm_dir` here?

**Review by Scott-Guest — APPROVED** [review]
Left a few comments, but LGTM overall. Thank you for doing this!

File: src/tests/integration/test_string.py (hunk 3 above)
Comment (tothtamas28) [reply src/tests/integration/test_string.py]: Done.

File: src/tests/integration/test_string.py (hunk 4 above)
Comment (tothtamas28) [reply src/tests/integration/test_string.py:219]: `KoreServer` is a wrapper for the Haskell backend's [`kore-rpc`](https://github.com/runtimeverification/haskell-backend/blob/master/docs/2022-07-18-JSON-RPC-Server-API.md) tool.
